Draft for the "0th" RFC (#11)
rust-lang/project-ffi-unwind: rfcs/0000-c-unwind-and-project-group-charter.md

- Feature Name: `simple_c_panic_abi`
- Start Date: 2019-08-29
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

* First, to modify Rust functions with the "C" abi to abort on unwinding, as proposed
  in [rust-lang/rust#52652].

> * Second, to introduce the "C unwind" ABI, with virtually all details

Reviewer: ```suggestion
* Second, to introduce the target-specific "C unwind" ABI (available on some targets only), with virtually all details
```

  (but not quite all) left as "to be determined".
    * The **intent** of this ABI, however, is to permit Rust panics to
      be "translated" into a "native" unwinding mechanism, though
      precisely which is dependent on the target platform or other

>       compiler options.

Reviewer: Which compiler options do you have in mind here ? 

_I think_ that the "native" unwinding mechanism is part of the target ABI, i.e., part of the target triple. So there should be no way to change it via compiler options, doing so requires a completely different target. 

For example, on windows, the mingw targets specify the unwinding ABI as part of their triple in the MinGW C toolchains `...-mingw-seh`, `...-mingw-sjlj`, etc. 

Currently, for Rust, the `x86_64-pc-windows-gnu` and `i686-pc-windows-gnu` use the SEH abi. If we wanted to use the SJLJ abi there, a different target triple, e.g., `x86_64-pc-windows-gnu-sjlj` feels like a better idea, because not only the Rust part of the ABI implementation has to change, but we need to call the C compilers and linkers with different flags, link different platform libraries that use this other ABI, etc. 

There is the potential use case of being able to link both C libraries that use SEH and C libraries that use SJLJ in such a platform, but then "C unwind" isn't enough, because it doesn't let you specify which ABI to use. Instead of maybe trying to make "C unwind" fit this use case, I think it would be better to leave solving this problem to some other language feature, e.g., adding two platform specific ABIs on that target only, like `"C SEH"` and `"C SJLJ"`, or something like that. That kind of ABI information needs to propagate correctly through function pointers and such.

So I would change the text above to:

> The **intent** of this ABI, however, is to permit Rust panics to
> be "translated" into the "native" unwinding mechanism of the platform,
> which is fixed by the platform target specification.

Note that this text is used below as well.

        * (In practice, this translation is a no-op, since Rust panics
          use the native mechanism, but this is explicitly not
          required for the future.)
    * Folks currently using the "C" ABI to do unwinding can migrate to this new
      "C unwind" ABI, which captures their intentions. Their code is no more stable
      than it was, but they are on the right path.
* Finally, to create a "project group" with the purpose of designing subsequent
  RFCs to flesh out the details of the "C unwind" ABI
    * The "project group" term is newly introduced, but it corresponds
      to a kind of working group -- one with the goal of fleshing out
      a particular proposal or completing a project.

>     * We aim to specify how "C unwind" works on major platforms

Reviewer: ```suggestion
    * We aim to specify how "C unwind" works on major platforms, adding support for new targets incrementally
```

    * Goal is to enable Rust panics to propagate across native frames
    * And perhaps to enable native exceptions to propagate across Rust frames

>     * But not to allow catching or throwing native exceptions from Rust code

Reviewer: ```suggestion
    * But we don't know at this point if the working group will end up allowing catching or throwing native exceptions from Rust code
```

@acfoltzer I think something like this is what you meant ?

them that could allow them to keep working (even if they are
continuing to rely on behavior that is not yet fully specified).

## Specific use cases

We have seen a number of unwinding use cases in the wild. This section
details some of those use cases and their specific requirements.

### Lucet: unwind past empty native frames

The **[Lucet]** compiler generates native frames that invoke Rust
helpers. These frames are always invoked from some Rust helper.  From
time to time, they may invoke other Rust helpers, resulting in native
frames "sandwiched" between two Rust frames:

[Lucet]: https://www.fastly.com/blog/announcing-lucet-fastly-native-webassembly-compiler-runtime

```
[Rust root frame]
     |
     v
[Native frames]
     |
     v
[Rust helper function]
```

This Rust helper function may panic. Lucet would like to be able to
catch this panic from the Rust root frame. They would like to have the
Native frames "gracefully unwind". However, the native frames
themselves do not contain any destructors or landing pads -- they
simply need to be "skipped over", in effect.

Moreover, Lucet executes only on a narrow range of platforms:

* XXX details

### mozjpeg: XXX

> XXX somebody fill this in XXX

Reviewer: ```suggestion
The `mozjpeg-rust` crate provides an idiomatic Rust wrapper over the `libjpeg` library. When this library errors, it calls a user-provided callback that is not allowed to return. The documentation mentions that this callback either aborts or `longjmp` somewhere. The `mozjpeg-rust` crates implements this callback to `panic!` instead, unwinding from the Rust callback into C, and then from C back into Rust.
```

### lua bindings and longjmp

Although longjmp/setjmp are not directly in scope for this group,
there is an interesting interaction that was uncovered and is worth
documenting. On Windows targets specifically, longjmp is implemented
using the same "Structured Exception Handling" (SEH) mechanism that is
used for C++ exceptions and other sorts of errors. As a result, in
some versions of Rust at least, we found that attempts to longjmp over
Rust frames triggered unwinding mechanisms. Note that in general a
longjmp over a Rust frame that contains destructors is [Undefined
Behavior] -- however, we would like to support longjmp for Rust frames
that do not contain destructors. At minimum, this group should take
the behavior of longjmp into account; if it is easy to do, we may also
introduce mechanisms or RFCs that help to specify the use of longjmp

> in particular cases.

Reviewer: So this isn't really that related to `longjmp`, although `longjmp` exhibits this issue. As the text says, on these platforms, these longjmps are implemented using "native unwinds". The problem is that on these platforms "native unwinds" can behave differently, depending on, e.g., how they are thrown. 

If this WG attacks the problem of how native unwinds behave in Rust, it would need to consider those potential differences in behavior. 

I'm not sure how to word that into this section which appears more focused on the issue discovered while using rlua on windows.

## Our proposal

This RFC proposes a multi-prong plan to solve this problem:

* First, to modify Rust functions with the "C" abi to abort on unwinding, as proposed
  in [rust-lang/rust#52652].

> * Second, to introduce the "C unwind" ABI, with virtually all details

Reviewer: ```suggestion
* Second, to introduce the target-specific "C unwind" ABI (available on some platforms only), with virtually all details
```

      be "translated" into a "native" unwinding mechanism, though
      precisely which is dependent on the target platform or other
      compiler options.
        * (In practice, this translation is a no-op, since Rust panics
          use the native mechanism, but this is explicitly not
          required for the future.)
    * Folks currently using the "C" ABI to do unwinding can migrate to this new
      "C unwind" ABI, which captures their intentions. Their code is no more stable
      than it was, but they are on the right path.
* Finally, to create a "project group" with the purpose of designing subsequent
  RFCs to flesh out the details of the "C unwind" ABI
    * The "project group" term is newly introduced, but it corresponds
      to a kind of working group -- one with the goal of fleshing out
      a particular proposal or completing a project.
    * We aim to specify how "C unwind" works on major platforms
    * Goal is to enable Rust panics to propagate across native frames
    * And perhaps to enable native exceptions to propagate across Rust frames
    * But not to allow catching or throwing native exceptions from Rust code

> The first two goals work together. We simul

Reviewer: ```suggestion
The first two goals work together.
```

>     * We aim to specify how "C unwind" works on major platforms

Reviewer: ```suggestion
    * We aim to specify how "C unwind" works on major platforms, adding support for new targets incrementally
```

The goal of introducing the "C unwind" ABI is simple: it allows
existing projects to migrate to the newer ABI and continue to work "as

> well as they ever did". In the meantime, we can tighten up the rules

Reviewer: ```suggestion
well as they ever did" or better, since they will avoid certain optimizations
that are incorrect for their use cases. In the meantime, we can tighten up the rules
```

around the "C" ABI.

Simply taking that step, however, does not address the root problem of
"de facto" dependencies. We've already seen that projects can and do
rely on unwinding.

The majority of this RFC is spent on describing the "C unwind" ABI and
its details. We do note a few details about the "project group", but we
expect the "project group" to largely define its own structure.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## The "C" ABI and unwinding

Functions declared with the "C" ABI are not permitted to unwind.  In
general, any attempt to unwind constitutes [Undefined Behavior], which

> means that the program may do arbitrary things.

Reviewer: ```suggestion
means that the program may do arbitrary things. Rust functions defined
with the "C" ABI are guaranteed to abort if an unwind attempts to
cross the function boundary.
```

Reviewer: That's mentioned below, but I think it is ok to be more optimistic here and repeat that this RFC removes the amount of UB in Rust by defining some of it to abort.

### Invoking foreign functions

This means, for example, that if you invoke a native "C" function and it
attempts to unwind (for example, because it invokes C++ code which contains a `throw`),
this result is [Undefined Behavior]:

```rust
extern "C" { fn fn_that_is_not_supposed_to_unwind(); }

fn foo() {
    // If `fn_that_is_not_supposed_to_unwind` should actually unwind, 
    // the result is undefined behavior.
    fn_that_is_not_supposed_to_unwind();
}
```

### Rust functions with C ABI

> In the case of Rust functions declared with the "C" ABI, attempts to

Reviewer: ```suggestion
In the case of Rust functions defined with the "C" ABI, attempts to
```

unwind over the function boundary will result in an abort:

```rust
extern "C" fn foo() {
    // This panic will trigger an unwind that tries to pass over an extern "C" 
    // boundary. It will trigger an abort.
    panic!();
}
```

## The "C unwind" ABI and unwinding

In contrast, functions declared with the "C unwind" ABI **are**
permitted to unwind. The precise unwinding mechanism in use will
depend on the target platform, but the intent is to match the
mechanism used by other systems programming language implementations,

> such as C++.

Reviewer: ```suggestion
such as C++ or C with `-fexceptions`.
```

fully specified fashion**. If you would like to learn more about the
specification effort, or to get involved, please visit the repository
for the [ffi-unwind project], which contains a lot more details.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Platform dependent behavior of "C" and "C unwind"

In general, the behavior of both the "C" and the "C unwind" ABIs is
highly dependent on the target triple and other switches that are
given to the Rust compiler. Therefore, there isn't a lot that can be
said that is fully independent of the platform.

## Relationship between "C" and "C unwind"

There is no a priori relationship between the "C" and "C unwind" ABIs.
However, on most platforms, it is expected that the "C" ABI would be
"compatible" with the "C unwind" ABI. In other words, consider some C
function `foo` that does not unwind: on most platforms, such a
function could safely be invoked with either the "C" or the "C unwind"
ABI. 

However, we do not guarantee such interop: this leaves room for
platforms where unwinding is transmitted through a special return
value or other such mechanism. (XXX do any such platforms exist? There

> is the C++ proposal)

Reviewer: So I'm not sure about these two paragraphs. I suppose that this a defense against "unknown unknowns" but I understand "C unwind" as "the C ABI + unwinding", so I kind of expect it to be equivalent to "C" when `-C panic=abort`, for example. 

The C++ proposal is compatible with the C ABI, and there is also a C proposal for it that adds support for the same C++ feature to C without breaking the current call ABI (I'm not sure if it breaks the unwind ABI on some platforms though, but since the C standard doesn't have a concept of unwinding, it breaks nothing there at least).
